### PR TITLE
Custom Flip Transition

### PR DESCRIPTION
**Pull request by caseycookie14**
* Changing our react-navigation to allow for the custom flip transition which is required from business
@sdg9 Did a lot of changes in this library, so he will have more insight to changes. Commit messages are detailed.

Changes in Card stack include:
* drawing non purged scenes for rendering each card in the card stack
* editing render to know when to hide scenes during flip
* masking white header when flipping by making it primary blue
* procesFlipAnimation method to hide

Changes in CardStackTransitioner include:
* determine if isFlipTranstion Prop
* pass onFlipStart, onFlipFromComplete, onFlipToComplete to Transitioner
* have to set transitionSpec.useNativeDriver = false, because there is a freeze issue with flipping transitions and some screens freeze
* pass isFlipFrom, isFlipTo, is FlipTransition to CardStack

Changes in ScenesReducer include: 
* reduxToComponentStateHelper defined to help sync up react-navigation and caribou

Changes in Transitioner include: 
* noViewableSceneChange logic is added
* start transition
* if we are a flip transition we split the transition into two
* else we go regular default transitions
* define getRegularAnimation and getFlipAnimation to help with transitions

### CODE REVIEW

File: src/views/CardSceneView.js [hunk 1]
```diff
@@ -19,38 +19,36 @@ type Props = {
...
-
   constructor(props: DFSCardNavigatorProps) {
     super(props);
     (this: any)._trackState = this._trackState.bind(this);
   }
 
   _trackState(route, data) {
     if (!route.trackingPageName) {
-      console.warn(`Attribute \`trackingPageName\` not defined in routes for ${route.routeName}`);
+      console.warn(
```
Comment (sdg9) [inline src/views/CardSceneView.js:29]: @caseycookie14 did we get the right branch?  Changes here aren't intentional are they?

Looks like we're losing some stuff.

Comment (caseycookie14) [reply src/views/CardSceneView.js:29]: @sdg9 yeah i think so, this change is in the first commit you made on your [flipAnimation2 branch](https://github.com/CaribouBot/react-navigation/pull/4/commits/f937af5007744d5e388f872520d816b1e607730c)

File: src/views/CardSceneView.js [hunk 2]
```diff
@@ -19,38 +19,36 @@ type Props = {
...
         {...routeProps}
         key={scene.key}
-
         // routeKey used for ScreenFocusAware
         routeKey={scene.route.key}
         handleNavigate={this.props.handleNavigate}
         trackPage={data => this._trackState(scene.route, data)}
-        handleBack={this.props.handleBack}
-        isLeftSplitPaneComponent={this.props.isLeftSplitPaneComponent}
+        handleBack={this.props.handleBackAction}
```
Comment (sdg9) [inline src/views/CardSceneView.js]: @caseycookie14 I know I added this but looks like a mistake, per #1 can you revert this change?

Also ` isLeftSplitPaneComponent={this.props.isLeftSplitPaneComponent}` should remain.

File: src/views/Transitioner.js [hunk 3]
```diff
@@ -105,24 +111,57 @@ class Transitioner extends React.Component<Props, State> {
       this.props.navigation.state
     );
 
-    if (nextScenes === this.state.scenes) {
+    const noSceneChange = nextScenes === this.state.scenes;
```
Comment (sdg9) [inline src/views/Transitioner.js]: We probably don't need any of this logic if we aren't doing the update redux without animation.  We had that problem where it froze hence we went the purge route.

That is pretty much everything in this file before line 215 `if (this.props.isFlipTransition) {` is probably not required.

Might be cleaner to not have it all if it's not being used.

File: src/views/ScenesReducer.js [hunk 4]
```diff
@@ -73,6 +73,16 @@ function areRoutesShallowEqual(
   return shallowEqual(one, two);
 }
 
+export function reduxToComponentStateHelper(nextState: NavigationState) {
```
Comment (sdg9) [inline src/views/ScenesReducer.js]: This really isn't being used since we are going to isPurged route instead of deleting items off the stack.

File: src/views/Transitioner.js [hunk 5]
```diff
@@ -293,6 +316,70 @@ function isSceneActive(scene: NavigationScene): boolean {
...
+  const flipToAnimation =
+    indexHasChanged && positionHasChanged
+      ? [
+          Animated.timing(progress, {
+            ...transitionSpec,
+            toValue: 1,
+          }),
+          Animated.timing(position, {
+            ...transitionSpec,
+            toValue: nextProps.navigation.state.index,
```
Comment (ramapriya2005) [inline src/views/Transitioner.js:373]: For all these animations, can we add below code to improve performance.
useNativeDriver: true

Comment (caseycookie14) [reply src/views/Transitioner.js:373]: @ramapriya2005 we cannot use useNativeDriver: true, because the app will freeze for an unknown reason

File: src/views/ScenesReducer.js (hunk 4 above)
Comment (caseycookie14) [reply src/views/ScenesReducer.js]: @sdg9 done

File: src/views/Transitioner.js (hunk 3 above)
Comment (caseycookie14) [reply src/views/Transitioner.js]: @sdg9 done

**Review by sdg9 — APPROVED** [review]
